ua/extension_object.go

```diff
@@ -5,17 +5,41 @@
 package ua
 
 import (
+	"fmt"
+	"reflect"
+
 	"github.com/gopcua/opcua/debug"
-	"github.com/gopcua/opcua/id"
 )
 
 // eotypes contains all known extension objects.
-var eotypes = NewTypeRegistry()
+var eotypes = NewFuncRegistry()
 
 // RegisterExtensionObject registers a new extension object type.
 // It panics if the type or the id is already registered.
 func RegisterExtensionObject(typeID *NodeID, v interface{}) {
-	if err := eotypes.Register(typeID, v); err != nil {
+	ef := func(vv interface{}) ([]byte, error) {
+		// TODO check/ensure vv is of type v ?
+		// TODO
+		return Encode(vv)
+	}
+	df := func(b []byte, vv interface{}) error {
+		rv := reflect.ValueOf(vv)
+		if rv.Kind() != reflect.Pointer || rv.IsNil() {
+			return fmt.Errorf("incorrect type to decode into")
+		}
+		r := reflect.New(reflect.TypeOf(v).Elem()).Interface()
+		buf := NewBuffer(b)
+		buf.ReadStruct(r)
+		reflect.Indirect(rv).Set(reflect.ValueOf(r))
+		return nil
+	}
+	RegisterExtensionObjectFunc(typeID, ef, df)
+}
+
+// RegisterExtensionObjectFunc registers a new extension object type using encode and decode functions
+// It panics if the type or the id is already registered.
+func RegisterExtensionObjectFunc(typeID *NodeID, ef encodefunc, df decodefunc) {
```

Comment:
If this method is public the `encodefunc` and `decodefunc` should also be public imo

Reply:
Why? Its essentially an interface that you can still adhere to from outside of the package (see read_unknow_node_id_test.go for example). I don't think anyone should even know about the type name, just the function handle